# Python для анализа данных

## Введение в Pandas 

Pandas - библиотека для работы с табличными данными в питоне.
* Документация: https://pandas.pydata.org/
* 10 minutes intro: https://pandas.pydata.org/pandas-docs/stable/user_guide/10min.html#min
* Pandas Cheat-Sheet: https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf

In [182]:
import pandas as pd # импортировали библиотеку pandas и назвали ее pd 

In [183]:
data = pd.read_csv('train.csv')

Функция read_csv читает данные из файла формата csv данные и преобразует в pandas.DataFrame. Аналогичная функция read_excel может читать данные в офрмате xls(x).

Посмотрим на наши данные:

In [184]:
data.head(10) # функция head() показывает первые строки датафрейма, по умолчанию 5

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [185]:
data.tail(10)

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

In [186]:
data.shape # функция shape показывает размерность датафрейма (строк, столбцов)

(891, 12)

In [187]:
data.columns # список столбцов 

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

Описание признаков из [источника данных](https://www.kaggle.com/competitions/titanic/data):

**PassengerId** - id пассажира

**Survived** бинарная переменная: выжил пассажирил (1) или нет (0)

**Pclass** - класс пассажира

**Name** - имя пассажира

**Sex** - пол пассажира

**Age** - возраст пассажира

**SibSp** - количество родственников (братьев, сестер, супругов) пассажира на борту

**Parch** - количество родственников (родителей / детей) пассажира на борту

**Ticket** - номер билета

**Fare** - тариф (стоимость билета)

**Cabin** - номер кабины

**Embarked** - порт, в котором пассажир сел на борт (C - Cherbourg, S - Southampton, Q = Queenstown)

Так можно обратиться к столбцу:

In [188]:
data['Age'].head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

In [189]:
data.Age.head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

In [190]:
data.age

AttributeError: 'DataFrame' object has no attribute 'age'

Или к нескольким столбцам сразу:

In [ ]:
data[['Age', 'Sex']].head()

A так - к строке по индексу:

https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html

In [ ]:
data.loc[3, ['Name', 'Age', 'Survived']]

In [ ]:
data.iloc[3, 3:5]

In [ ]:
data.loc[0]

In [ ]:
data.iloc[0]

In [ ]:
data.iloc[1:3] # строки с 1 по 3

In [ ]:
data.dtypes
# object - имеется в виду строчной тип данных 

In [ ]:
data.describe()

data.describe().applymap('{:,.2f}'.format) # чтобы сделать форматирование читабельнее
# в расчете пропущенные ячейки не учитываются

In [ ]:
data[['Age', 'Fare', 'Survived']].describe() # также можно применять только к отдельным колонкам

Статистики:

* count - количество непустых значений
* mean - среднее значение
* std - стандартное отклонение – мера разброса в наборе числовых данных. Выражаясь простыми словами, насколько далеко от среднего арифметического (mean) находятся точки данных. Чем выше значение стандартного отклонения, тем более разбросаны значения в выборке. И наоборот, чем ниже значение стандартного отклонения, тем более плотно упакованы значения.
* min - минимальное значение
* 25% - 25% перцентиль
* 50% - 50% перцентиль
* 75% - 75% перцентиль
* max - максимальное значение

Подробнее про квантили и перцентили: https://vk.com/@hsotalks-zachem-yaschiku-usy-kvartili-raspredeleniya

In [ ]:
data.describe(include=['object'])

In [ ]:
data['Age'].min()

In [ ]:
data['Age'].max()

In [ ]:
data['Age'].median()

In [ ]:
data['Age'].quantile(0.9)

In [ ]:
data['Age'].mean()

In [ ]:
data['Age'] > data['Age'].mean()

In [ ]:
data[data['Age'] > data['Age'].mean()]

In [ ]:
data[data['Age'] > data['Age'].mean()]['Name'].head()

In [ ]:
data[(data['Age'] > data['Age'].mean()) & (data['Pclass'] == 1)]

In [ ]:
data[data['Fare'] > 200] #.shape[0]

In [ ]:
q = (data['Age'] > data['Age'].mean()) & (data['Pclass'] == 1) & \
    (data['Sex'] == 'female')

In [ ]:
data[q].shape
# 55- кол-во строк, 12 - кол-во колонок

In [ ]:
data[q]

In [ ]:
data[(data['Sex'] == 'female') | ((data['Pclass'] == 1))]

In [ ]:
data[data['PassengerId'] == 2][['Name','Pclass']]

In [ ]:
loc работает с индексами(которые могут быть слова)
iloc работает с относительными индексами, т.е. только циферки от 0 ..

In [ ]:
data.loc[1:3, 'Survived':'Sex'] # строки с 1 по 3, колонки от Survived до Sex

In [ ]:
data.iloc[1:3, 1:5]

Кроме того, можно выбирать объекты, удовлетворяющие каким-то свойствам, например, все пассажиры-женщины:

In [ ]:
data[data.Sex == 'female'].head()

Пассажиры первого класса:

In [ ]:
data[(data.Pclass == 1) & (data.Sex == 'female')].shape

Пассажиры первого или второго классов: # (равно предыдущему выражению)

In [ ]:
data[data.Pclass.isin([1,2])].head()

In [ ]:
data.Pclass.isin([1, 2])

Пассажиры младше 18:

In [ ]:
data[data.Age < 18].shape

Девушки в возрасте от 18 до 25, путешествующие в одиночку (без каких-либо родственников):

In [ ]:
data[(data.Sex == 'female') & (data.Age > 18) & (
    data.Age < 25) & (data.SibSp == 0) & (data.Parch == 0)]

Сколько таких путешественниц?

In [ ]:
data[(data.Sex == 'female') & (data.Age > 18) & (data.Age < 25)
     & (data.SibSp == 0) & (data.Parch == 0) & (data.Survived == 0)].shape[0]

<hr>

### Задачи

1) Посчитайте количество пассажиров первого класса, которые сели на борт в Саутгемптоне.

In [ ]:
data

In [ ]:
# TODO
data[(data['Pclass'] == 1) & (data['Embarked'] == 'S')].shape[0]

2) Сколько пассажиров третьего класса, которые путешествовали в компании 2 или более родственников?

In [ ]:
# TODO
data[(data['Pclass'] == 3) & (data['SibSp'] + data['Parch'] >= 2) ] #.shape

3) Сколько в среднем стоил билет первого класса?

In [ ]:
# TODO
round(data[(data['Pclass'] == 1)]['Fare'].mean(), 2)

<hr>

Иногда нужно создать новый признак из уже существующих, например, нам интересно, сколько всего родственников путешествовало с каждым пассажиром - просто сложим столбцы SibSp и Parch и поместим сумму в новый столбец FamilySize. Такая процедура называет broadcasting. 

In [ ]:
data['FamilySize'] = data['SibSp'] + data['Parch']
data.head()

In [ ]:
data.sample(5) #рандомный набор

А теперь давайте создадим переменную, которая бы нам показывала, что пассажир ехал в одиночку. Такой пассажир путешествовал без родственников. Мы напишем условие с помощью анонимной функции (1, если FamilySize равно 0 и 0 во всех остальных случаях) и применим ее к столбцу FamilySize с помощью метода .apply().

In [ ]:
data['Alone'] = data['FamilySize'].apply(lambda x: 1 if x == 0 else 0)
data.head()

In [ ]:
data[data['Alone'] == 0]

Функция, которую применяем к столбцу, может быть и посложнее. Давайте из каждого имени пассажира достанем его титул. Сначала потренируемся на одном пассажире.

In [ ]:
data.loc[0]['Name']

Ок, выбрали имя. Это строка. Давайте подумаем, как достать из нее титул. Вроде бы титул всегда после фамилии. Разобьем строку по пробелу и доставим второй (первый по индексу) элемент.

In [ ]:
data.loc[0]['Name'].split(', ')[1].split('.')[0]

Ура! Теперь напишем функцию, которая будет забирать титул из имени, а потом применим ее к колонке Name.

In [209]:
def return_title(full_name):
    return full_name.split(', ')[1].split('.')[0] + '.'

Теперь сформируем новый столбец family_name из столбца Name с помощью написанной нами функции:

In [210]:
data['Title'] = data.Name.apply(return_title)

In [211]:
data.sample(5)

PassengerId  Survived  Pclass  \
805          806         0       3   
284          285         0       1   
26            27         0       3   
617          618         0       3   
29            30         0       3   

                                                Name     Sex   Age  SibSp  \
805                        Johansson, Mr. Karl Johan    male  31.0      0   
284                       Smith, Mr. Richard William    male   NaN      0   
26                           Emir, Mr. Farred Chehab    male   NaN      0   
617  Lobb, Mrs. William Arthur (Cordelia K Stanlick)  female  26.0      1   
29                               Todoroff, Mr. Lalio    male   NaN      0   

     Parch     Ticket     Fare Cabin Embarked  Age_Median  Age_Median_Sex  \
805      0     347063   7.7750   NaN        S        31.0            31.0   
284      0     113056  26.0000   A19        S        28.0            29.0   
26       0       2631   7.2250   NaN        C        28.0            29.0   
617      0  A/5. 3336  16.1000   NaN        S        26.0            26.0   
29       0     349216   7.8958   NaN        S        28.0            29.0   

    Title  
805   Mr.  
284   Mr.  
26    Mr.  
617  Mrs.  
29    Mr.

Кстати, удалить колонку можно так. В нашем анализе мы не будем использовать колонку Ticket, даайте удалим ее.

In [ ]:
del data['Ticket']

In [ ]:
data['Title'].unique()

In [ ]:
data['Title'].nunique()

Полезно посмотреть, какие значения содержатся в столбце. Для категориальных данных можно посчитать частоту встречаемости каждого значения с помошью функции value_counts:

In [ ]:
data['Title'].value_counts()

In [ ]:
data['Title'].value_counts(normalize=True) # * 100

Очень много уникальных значений! Обычно это не очень хорошо для статистического анализа, давайте все такие титулы переиминуем в Misc (другие).

In [235]:
data['Title'] = data['Title'].apply(lambda x: 'Misc' if x not in ['Mr.', 'Miss.', 'Mrs.', 'Master.'] else x)

In [ ]:
data['Title'].value_counts()

In [ ]:
data['Pclass'].value_counts()

In [ ]:
data['Embarked'].value_counts()

Данные можно сортировать:

In [ ]:
data.sort_values(by=['Age']).head() # сортируем по возрасту, по умолчанию сортирвка по возрастанию

In [ ]:
data.sort_values(by=['Age']).iloc[0]
data.sort_values(by=['Age']).iloc[803]

In [ ]:
data.sort_values(by=['Age'], ascending=False).head() # сортируем по возрасту, теперь по убыванию

In [ ]:
data.sort_values(
    by=['Age', 'Fare'],
    ascending=False).head(15)  # сортируем сперва по возрасту (по убыванию),
# потом стоимости билета  (по убыванию)

In [ ]:
data.sort_values(by=['Age', 'Fare'], ascending=[False, True]).head(15)
# сортируем сперва по возрасту (по убыванию),
# потом стоимости билета  (по возрастанию)

И группировать:

In [ ]:
data.groupby('Pclass')['Fare'].mean()

In [ ]:
data.groupby('Pclass')[['Fare']].mean() # так просто симпатичней выведет

In [ ]:
data.groupby('Pclass')[['Fare']].mean().shape

In [ ]:
data.groupby('Pclass')[['Fare']].mean().index

In [ ]:
data.groupby('Pclass')[['Fare']].mean().reset_index()

In [ ]:
data.groupby('Pclass')['Fare'].min()

In [ ]:
data[data['Fare'] == 0]

In [ ]:
data.groupby('Pclass')['Fare'].max()

In [ ]:
data.groupby('Pclass')['Fare'].agg(['min', 'max', 'median', 'mean'])

In [ ]:
data.groupby('Sex') # разбиение всех объектов на 2 группы по полу - возращает просто сгруппированый датафрейм

In [ ]:
# группируем по полу и считаем для каждого пассажирова разных классов
data.groupby('Sex')['Pclass'].value_counts()

In [ ]:
data.groupby('Sex')['Age'].mean() # средний возраст для пассажиров каждого из полов

## И еще чуть-чуть! Работа с пропущенными значениями.

Если вы помните, то переменная Age содержит пропущенные значения. Давайте посмотрим информацию об этой колонке.


In [ ]:
data[['Age']].info()  

# обратите внимание - двойные скобки. Так pandas нам вернет датафрейм из одной колонки, а не список.
# А метод info() работает только с датафреймом

In [ ]:
data.shape

In [ ]:
type(data['Age'])  # вот так - объект подобный списку (Series)

In [191]:
type(data[['Age']])  # а вот так - датафрейм

pandas.core.frame.DataFrame

Вернемся к info(). Мы видим, что из 891 наблюдения у нас только 714 ненулевых значений. Значит, у этих пассажиров возраст неизвестен. Ваши идеи, что можно с этим сделать?

Есть разные варианты того, что делать с пропущенными значениями - от "не делать ничего и выкинуть их" до "давайте предскажем их значения с помощью нейронки". Почитать можно здесь: 

https://towardsdatascience.com/6-different-ways-to-compensate-for-missing-values-data-imputation-with-examples-6022d9ca0779

Мы с вами попробуем сделать второй по сложности вариант (после не делать ничего) и заменить их средним значением (средним или медианой). Для категориальных данных можно заполнять пропущенные значения модой.

Пропущенные значения могут быть закодированы по-разному - 0, 'No response', '999'. В итоге мы их всегда хотим привести к объекту NaN (not a number), с которым могут работать методы pandas. В нашем датасете они уже нужного формата. В других случаях, нужно будет отфильтровать значения и привести их к нужному виду.


In [192]:
print(data.loc[5, 'Age'])
print(type(data.loc[5, 'Age']))

nan
<class 'numpy.float64'>


In [193]:
data[data['Age'].isnull()].head() 
# выводим значения датафрейма, в которых отсутствует возраст
# Они возращают True методу .isnull()

PassengerId  Survived  Pclass                           Name     Sex  Age  \
5             6         0       3               Moran, Mr. James    male  NaN   
17           18         1       2   Williams, Mr. Charles Eugene    male  NaN   
19           20         1       3        Masselmani, Mrs. Fatima  female  NaN   
26           27         0       3        Emir, Mr. Farred Chehab    male  NaN   
28           29         1       3  O'Dwyer, Miss. Ellen "Nellie"  female  NaN   

    SibSp  Parch  Ticket     Fare Cabin Embarked  
5       0      0  330877   8.4583   NaN        Q  
17      0      0  244373  13.0000   NaN        S  
19      0      0    2649   7.2250   NaN        C  
26      0      0    2631   7.2250   NaN        C  
28      0      0  330959   7.8792   NaN        Q

In [194]:
data['Age'].median() # вспомним какая у нас медиана

28.0

In [195]:
data['Age_Median'] = data['Age'].fillna(data['Age'].median())   #fillna сделает расчет только для полей которые Nan
# сохраняю результат заполнения в новую колонку

In [196]:
data[data['Age'].isnull()].head(10)  
# смотрим, что произошло с возрастом в новой колонке у тех, у кого он отсутсвовал

PassengerId  Survived  Pclass  \
5             6         0       3   
17           18         1       2   
19           20         1       3   
26           27         0       3   
28           29         1       3   
29           30         0       3   
31           32         1       1   
32           33         1       3   
36           37         1       3   
42           43         0       3   

                                              Name     Sex  Age  SibSp  Parch  \
5                                 Moran, Mr. James    male  NaN      0      0   
17                    Williams, Mr. Charles Eugene    male  NaN      0      0   
19                         Masselmani, Mrs. Fatima  female  NaN      0      0   
26                         Emir, Mr. Farred Chehab    male  NaN      0      0   
28                   O'Dwyer, Miss. Ellen "Nellie"  female  NaN      0      0   
29                             Todoroff, Mr. Lalio    male  NaN      0      0   
31  Spencer, Mrs. William Augustus (Marie Eugenie)  female  NaN      1      0   
32                        Glynn, Miss. Mary Agatha  female  NaN      0      0   
36                                Mamee, Mr. Hanna    male  NaN      0      0   
42                             Kraeff, Mr. Theodor    male  NaN      0      0   

      Ticket      Fare Cabin Embarked  Age_Median  
5     330877    8.4583   NaN        Q        28.0  
17    244373   13.0000   NaN        S        28.0  
19      2649    7.2250   NaN        C        28.0  
26      2631    7.2250   NaN        C        28.0  
28    330959    7.8792   NaN        Q        28.0  
29    349216    7.8958   NaN        S        28.0  
31  PC 17569  146.5208   B78        C        28.0  
32    335677    7.7500   NaN        Q        28.0  
36      2677    7.2292   NaN        C        28.0  
42    349253    7.8958   NaN        C        28.0

In [197]:
data.head() # А у всех остальных - их нормальный возраст.

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_Median  
0      0         A/5 21171   7.2500   NaN        S        22.0  
1      0          PC 17599  71.2833   C85        C        38.0  
2      0  STON/O2. 3101282   7.9250   NaN        S        26.0  
3      0            113803  53.1000  C123        S        35.0  
4      0            373450   8.0500   NaN        S        35.0

Такой метод один из самых топорных, но сойдет для первого знакомства или черновика анализа в серьезном исследовании. Давайте попробуем теперь немного улучшить его. Посмотрим, как отличается медианный возраст для мужчин и женщин.

In [198]:
data.groupby('Sex')['Age'].median()

Sex
female    27.0
male      29.0
Name: Age, dtype: float64

In [199]:
data.groupby('Sex')["Age"].transform('median')

0      29.0
1      27.0
2      27.0
3      27.0
4      29.0
       ... 
886    29.0
887    27.0
888    27.0
889    29.0
890    29.0
Name: Age, Length: 891, dtype: float64

Разница два года! Было бы логично в наших данных заполнять недостающие значения по полу.

In [200]:
data["Age_Median_Sex"] = data["Age"].fillna(data.groupby('Sex')["Age"]\
                                            .transform('median'))

In [201]:
data[data['Age'].isnull()].head() 

PassengerId  Survived  Pclass                           Name     Sex  Age  \
5             6         0       3               Moran, Mr. James    male  NaN   
17           18         1       2   Williams, Mr. Charles Eugene    male  NaN   
19           20         1       3        Masselmani, Mrs. Fatima  female  NaN   
26           27         0       3        Emir, Mr. Farred Chehab    male  NaN   
28           29         1       3  O'Dwyer, Miss. Ellen "Nellie"  female  NaN   

    SibSp  Parch  Ticket     Fare Cabin Embarked  Age_Median  Age_Median_Sex  
5       0      0  330877   8.4583   NaN        Q        28.0            29.0  
17      0      0  244373  13.0000   NaN        S        28.0            29.0  
19      0      0    2649   7.2250   NaN        C        28.0            27.0  
26      0      0    2631   7.2250   NaN        C        28.0            29.0  
28      0      0  330959   7.8792   NaN        Q        28.0            27.0

In [202]:
data.head(6) # Опять проверяем, что это все применилось только к нужным людям

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   

   Parch            Ticket     Fare Cabin Embarked  Age_Median  Age_Median_Sex  
0      0         A/5 21171   7.2500   NaN        S        22.0            22.0  
1      0          PC 17599  71.2833   C85        C        38.0            38.0  
2      0  STON/O2. 3101282   7.9250   NaN        S        26.0            26.0  
3      0            113803  53.1000  C123        S        35.0            35.0  
4      0            373450   8.0500   NaN        S        35.0            35.0  
5      0            330877   8.4583   NaN        Q        28.0            29.0

Разберем как работает предыдущий кусок кода

In [203]:
# эта функция возвращает нам колонку возраст, где все значения заменены медианой по условию пола
# data.groupby('Sex') - группирует наши значения по полу
# ['Age'] - колонка, с которой работаем
# transform('median') - высчитывает медианный возраст для каждого пола и подставляет ее вместо значения

data.groupby('Sex')['Age'].transform('median').head()

0    29.0
1    27.0
2    27.0
3    27.0
4    29.0
Name: Age, dtype: float64

In [179]:
data.iloc[5]

PassengerId                      6
Survived                         0
Pclass                           3
Name              Moran, Mr. James
Sex                           male
Age                            NaN
SibSp                            0
Parch                            0
Ticket                      330877
Fare                        8.4583
Cabin                          NaN
Embarked                         Q
Age_Median                    28.0
Age_Median_Sex                29.0
Name: 5, dtype: object

In [180]:
# data['Age'] = data['Age'].fillna(data['Age_Median_Sex'])

In [181]:
data.iloc[5]

PassengerId                      6
Survived                         0
Pclass                           3
Name              Moran, Mr. James
Sex                           male
Age                           29.0
SibSp                            0
Parch                            0
Ticket                      330877
Fare                        8.4583
Cabin                          NaN
Embarked                         Q
Age_Median                    28.0
Age_Median_Sex                29.0
Name: 5, dtype: object

<hr>

## Задание
Заполните осутствующие значения переменной "возраст" на основании титула.

In [206]:
data.Age.isna().sum()

177

In [213]:
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Age_Median  \
0        0         A/5 21171   7.2500   NaN        S        22.0   
1        0          PC 17599  71.2833   C85        C        38.0   
2        0  STON/O2. 3101282   7.9250   NaN        S        26.0   
3        0            113803  53.1000  C123        S        35.0   
4        0            373450   8.0500   NaN        S        35.0   
..     ...               ...      ...   ...      ...         ...   
886      0            211536  13.0000   NaN        S        27.0   
887      0            112053  30.0000   B42        S        19.0   
888      2        W./C. 6607  23.4500   NaN        S        28.0   
889      0            111369  30.0000  C148        C        26.0   
890      0            370376   7.7500   NaN        Q        32.0   

     Age_Median_Sex  Title  
0              22.0    Mr.  
1              38.0   Mrs.  
2              26.0  Miss.  
3              35.0   Mrs.  
4              35.0    Mr.  
..              ...    ...  
886            27.0   Rev.  
887            19.0  Miss.  
888            27.0  Miss.  
889            26.0    Mr.  
890            32.0    Mr.  

[891 rows x 15 columns]

In [236]:
data.groupby('Title')["Age"].median()

Title
Master.     3.5
Misc       44.5
Miss.      21.0
Mr.        30.0
Mrs.       35.0
Name: Age, dtype: float64

In [240]:
# TODO
# data["MedianAge"] = 
data["MedianAge"] = data['Age'].fillna(data.groupby('Title')["Age"].transform('median'))
#                                             .transform('median')
data.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  Age_Median  \
0      0         A/5 21171   7.2500   NaN        S        22.0   
1      0          PC 17599  71.2833   C85        C        38.0   
2      0  STON/O2. 3101282   7.9250   NaN        S        26.0   
3      0            113803  53.1000  C123        S        35.0   
4      0            373450   8.0500   NaN        S        35.0   
5      0            330877   8.4583   NaN        Q        28.0   
6      0             17463  51.8625   E46        S        54.0   
7      1            349909  21.0750   NaN        S         2.0   
8      2            347742  11.1333   NaN        S        27.0   
9      0            237736  30.0708   NaN        C        14.0   

   Age_Median_Sex    Title  MedianAge  
0            22.0      Mr.       22.0  
1            38.0     Mrs.       38.0  
2            26.0    Miss.       26.0  
3            35.0     Mrs.       35.0  
4            35.0      Mr.       35.0  
5            29.0      Mr.       30.0  
6            54.0      Mr.       54.0  
7             2.0  Master.        2.0  
8            27.0     Mrs.       27.0  
9            14.0     Mrs.       14.0

<hr>

### Заполнение по моде для категориальных переменных

Тоже самое (почти!) работает и для категориальных переменных.

In [241]:
data[data["Embarked"].isnull()]

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  Age_Median  \
61   female  38.0      0      0  113572  80.0   B28      NaN        38.0   
829  female  62.0      0      0  113572  80.0   B28      NaN        62.0   

     Age_Median_Sex  Title  MedianAge  
61             38.0  Miss.       38.0  
829            62.0   Mrs.       62.0

Давайте посмотрим, что возвращает нам функция мода - не число, как например median или mean, а список. 

In [242]:
data['Embarked'].mode()

0    S
Name: Embarked, dtype: object

Чтобы передать ее результат методу fillna, нам нужно "вытащить" значение из него (а это мы умеем делать - оно лежит под нулевым индексом).

In [243]:
data['Embarked'].mode()[0]

'S'

In [244]:
# применяем
data["Embarked_Mode"] = data["Embarked"].fillna(data['Embarked'].mode()[0])

In [245]:
# проверяем
data.loc[61]

PassengerId                        62
Survived                            1
Pclass                              1
Name              Icard, Miss. Amelie
Sex                            female
Age                              38.0
SibSp                               0
Parch                               0
Ticket                         113572
Fare                             80.0
Cabin                             B28
Embarked                          NaN
Age_Median                       38.0
Age_Median_Sex                   38.0
Title                           Miss.
MedianAge                        38.0
Embarked_Mode                       S
Name: 61, dtype: object

In [246]:
data[data["Embarked"].isnull()]

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  Age_Median  \
61   female  38.0      0      0  113572  80.0   B28      NaN        38.0   
829  female  62.0      0      0  113572  80.0   B28      NaN        62.0   

     Age_Median_Sex  Title  MedianAge Embarked_Mode  
61             38.0  Miss.       38.0             S  
829            62.0   Mrs.       62.0             S

### Самое важное! Сохраним результаты изменений таблицы

In [251]:
data.to_csv('titanic_new.csv', index = False)
# data.to_csv('titanic_new.csv')

In [250]:
df = pd.read_csv('titanic_new.csv')
df.head(6)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   

   Parch            Ticket     Fare Cabin Embarked  Age_Median  \
0      0         A/5 21171   7.2500   NaN        S        22.0   
1      0          PC 17599  71.2833   C85        C        38.0   
2      0  STON/O2. 3101282   7.9250   NaN        S        26.0   
3      0            113803  53.1000  C123        S        35.0   
4      0            373450   8.0500   NaN        S        35.0   
5      0            330877   8.4583   NaN        Q        28.0   

   Age_Median_Sex  Title  MedianAge Embarked_Mode  
0            22.0    Mr.       22.0             S  
1            38.0   Mrs.       38.0             C  
2            26.0  Miss.       26.0             S  
3            35.0   Mrs.       35.0             S  
4            35.0    Mr.       35.0             S  
5            29.0    Mr.       30.0             Q